pallets/parachain-staking/src/lib.rs

```diff
@@ -1451,215 +2161,175 @@ pub mod pallet {
 					Self::update_active(collator.clone(), new_total);
 				}
 			}
 			let new_total_locked = <Total<T>>::get() + amount;
 			<Total<T>>::put(new_total_locked);
-			<CollatorState2<T>>::insert(&collator, state);
-			<NominatorState2<T>>::insert(&acc, nominator);
-			Self::deposit_event(Event::Nomination(acc, amount, collator, nominator_position));
+			<CandidateState<T>>::insert(&collator, state);
+			<DelegatorState<T>>::insert(&acc, delegator_state);
+			Self::deposit_event(Event::Nomination(acc, amount, collator, delegator_position));
 			Ok(().into())
 		}
-		/// Request to leave the set of nominators. If successful, the nominator is scheduled
-		/// to exit
-		#[pallet::weight(<T as Config>::WeightInfo::leave_nominators(*nomination_count))]
-		pub fn leave_nominators(
+		#[pallet::weight(0)]
+		/// Request to leave the set of delegators. If successful, the caller is scheduled
+		/// to be allowed to exit. Success forbids future delegator actions until the request is
+		/// invoked or cancelled.
+		pub fn leave_delegators(
 			origin: OriginFor<T>,
 			nomination_count: u32,
 		) -> DispatchResultWithPostInfo {
 			let acc = ensure_signed(origin)?;
-			let mut state = <NominatorState2<T>>::get(&acc).ok_or(Error::<T>::NominatorDNE)?;
+			let mut state = <DelegatorState<T>>::get(&acc).ok_or(Error::<T>::NominatorDNE)?;
 			ensure!(!state.is_leaving(), Error::<T>::NominatorAlreadyLeaving);
 			ensure!(
-				nomination_count >= (state.nominations.0.len() as u32),
+				nomination_count >= (state.delegations.0.len() as u32),
 				Error::<T>::TooLowNominationCountToLeaveNominators
 			);
-			let mut exits = <ExitQueue2<T>>::get();
-			let now = <Round<T>>::get().current;
-			let when = now + T::LeaveNominatorsDelay::get();
-			exits.schedule_nominator_exit::<T>(acc.clone(), when)?;
-			state.leave(when);
-			state.scheduled_revocations_total = state.total;
-			state.scheduled_revocations_count = state.nominations.0.len() as u32;
-			<ExitQueue2<T>>::put(exits);
-			<NominatorState2<T>>::insert(&acc, state);
+			let (now, when) = state.leave::<T>();
+			<DelegatorState<T>>::insert(&acc, state);
 			Self::deposit_event(Event::NominatorExitScheduled(now, acc, when));
 			Ok(().into())
 		}
-		/// Request to revoke an existing nomination. If successful, the nomination is scheduled
-		/// to exit
+		#[pallet::weight(0)]
+		/// Execute the right to exit the set of delegators and revoke all ongoing delegations.
+		pub fn execute_leave_delegators(
+			origin: OriginFor<T>,
+			delegator: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+			let state = <DelegatorState<T>>::get(&delegator).ok_or(Error::<T>::NominatorDNE)?;
+			ensure!(state.can_leave::<T>()?, Error::<T>::NominatorCannotLeaveYet);
+			for bond in state.delegations.0 {
+				if let Err(error) =
+					Self::delegator_leaves_collator(delegator.clone(), bond.owner.clone())
+				{
+					log::warn!("Nominator exit collator failed with error: {:?}", error);
```

Comment:
So if there is an error here is likely that we were not able to unreserve, and we just log a warning and remove the delegator with some delegation still going on. Am I right on this?

Reply:
`unreserve` is infallible and cannot error.

There are 2 possible errors that can be emitted from `delegator_leave_collator`: (1) the candidate state does not exist (2) the delegation cannot be found in the candidate state. The code enforces the assumption that neither of these errors are possible if a delegation exists in the delegator state. If this is ever violated, it is a fatal error.

Reply:
I updated the warn message to communicate that critical assumptions have been violated.